src/docfx/lib/log/Log.cs

```diff
@@ -0,0 +1,142 @@
+        }
+
+        public static void Progress(int done, int total)
+        {
+            try
+            {
+                Debug.Assert(t_scope.Value != null);
+
+                var now = DateTime.UtcNow;
+                if (done != total && now - s_lastProgressTime < TimeSpan.FromSeconds(1))
+                {
+                    return;
+                }
+                s_lastProgressTime = now;
+
+                var scope = t_scope.Value.Peek();
+                var percent = Math.Min(1.0, done / Math.Max(1.0, total));
+                var head = $"{scope.Name} ";
+                var tail = $" {(int)(percent * 100)}%, {done}/{total} [{ElapsedTime(scope.StartTime)}]";
+
+                var width = Math.Min(40, Console.BufferWidth - 1 - head.Length - tail.Length - 2);
+                var n = (int)(percent * width);
+                var progress = width > 0 ? $"[{new string('#', n)}{new string(' ', Math.Max(0, width - n))}]" : "";
+                var line = head + progress + tail;
+
+                WriteProgress(scope.Name, line);
+
+                scope.HasProgress = true;
+            }
+            catch
+            {
+                // Swallow console exceptions
+            }
+        }
+
+        public static void WriteError(ErrorLevel level, Error error)
+        {
+            try
+            {
+                lock (s_consoleLock)
+                {
+                    s_lastProgressName = null;
+
+                    if (!string.IsNullOrEmpty(error.File))
+                    {
+                        Console.ForegroundColor = ConsoleColor.DarkGray;
+                        Console.BackgroundColor = ConsoleColor.Black;
+                        Console.Write(error.File + ":");
+                        Console.ResetColor();
+                        Console.WriteLine();
+                    }
+
+                    Console.ForegroundColor = GetColor(level);
+                    Console.Write(error.Code + " ");
+                    Console.ForegroundColor = ConsoleColor.Gray;
+                    Console.WriteLine(error.Message);
+                    Console.ResetColor();
+                }
+            }
+            catch
+            {
+                // Swallow console exceptions
+            }
+        }
+
+        private static ConsoleColor GetColor(ErrorLevel level)
+        {
+            switch (level)
+            {
+                case ErrorLevel.Error:
+                    return ConsoleColor.Red;
+                case ErrorLevel.Warning:
+                    return ConsoleColor.Yellow;
+                default:
+                    return ConsoleColor.Cyan;
+            }
+        }
+
+        private static void WriteProgress(string name, string line)
+        {
+            lock (s_consoleLock)
+            {
+                if (s_lastProgressName == name)
+                {
+                    Console.SetCursorPosition(0, Console.CursorTop - 1);
```

Comment:
>SetCursorPosition [](start = 28, length = 17)

it's not supported in dotnet core?

Reply:
supported.

---
In reply to: [198068425](https://github.com/dotnet/docfx/pull/2942#discussion_r198068425) [](ancestors = 198068425)